### PR TITLE
Attachment management

### PR DESCRIPTION
**Pull request by konraddysput**
# Why

This diff adds attachment support to the library. In the past, the client's option allowed us to set up some attachments and we even had a logic to upload them, however:
* in the browser library, the attachment support allowed to a setup file path that was unsupported in the HTTP client
* the SDK client couldn't merge all attachments

This diff adds more flexibility around attachments. We can still define super simple API for our users and convert data to the model we expect to use/have. 

Once we finish the attachment support we can start working on the breadcrumb support

### CODE REVIEW

File: packages/node/src/builder/BacktraceClientBuilder.ts [hunk 1]
```diff
@@ -24,7 +26,21 @@ export class BacktraceClientBuilder extends BacktraceCoreClientBuilder<Backtrace
         this.attributeProviders.push(provider);
     }
 
+    /**
+     * Transform client attachments into the attachment model.
+     * @returns attachments
+     */
+    private transformAttachments(): BacktraceAttachment[] {
+        return (
+            this._options.attachments?.map((n) => (typeof n === 'string' ? new BacktraceFileAttachment(n) : n)) ?? []
```
[the remark this answers is not in the document]
Comment (konraddysput) [reply packages/node/src/builder/BacktraceClientBuilder.ts:35]: they can, but they can also pass './path/to/the/file' and we will convert it for them. The reason why we have it is to simplify the interface needed to start working with our library. The user needs to have minimal knowledge to do everything they want in node.

File: packages/node/src/model/BacktraceFileAttachment.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    constructor(private readonly _filePath: string) {
+        this.name = path.basename(this._filePath);
+    }
+
+    public get(): fs.ReadStream | undefined {
+        if (!fs.existsSync(this._filePath)) {
+            return undefined;
+        }
+        return fs.createReadStream(this._filePath);
+    }
```
[the remark this answers is not in the document]
Comment (konraddysput) [reply packages/node/src/model/BacktraceFileAttachment.ts:16]: Can you imagine downloading attachments with such content from Backtrace? It also consumes disk space.

File: packages/node/src/model/BacktraceFileAttachment.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import { BacktraceAttachment } from '@backtrace/sdk-core';
+import fs from 'fs';
+import path from 'path';
+
+export class BacktraceFileAttachment implements BacktraceAttachment<fs.ReadStream> {
```
[the remark this answers is not in the document]
Comment (konraddysput) [reply packages/node/src/model/BacktraceFileAttachment.ts:5]: We can rename it to BacktraceStreamAttachment, but the main purpose of this class is to let the user know we can just stream attachment. How? It's the implementation details. As a user I just want to pass it and trust the library. So is it worth it? I think BacktraceFileAttachment is simple enough and straightforward for everyone

File: packages/node/src/builder/BacktraceClientBuilder.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (perf2711) [reply packages/node/src/builder/BacktraceClientBuilder.ts:35]: But then they cannot pass any strings as attachments.

File: packages/node/src/BacktraceClient.ts [hunk 4]
```diff
@@ -1,15 +1,16 @@
...
     BacktraceRequestHandler,
 } from '@backtrace/sdk-core';
 import { AGENT } from './agentDefinition';
+import { BacktraceConfiguration } from './BacktraceConfiguration';
 import { BacktraceClientBuilder } from './builder/BacktraceClientBuilder';
 
 export class BacktraceClient extends BacktraceCoreClient {
     constructor(
-        options: BacktraceConfiguration,
+        options: CoreConfiguration,
```
[the remark this answers is not in the document]
Comment (konraddysput) [reply packages/node/src/BacktraceClient.ts:13]: Because there is a type mismatch between BacktraceConfiguration that someone in Node wants to configure and the definition the core library expects. 

tldr:

the core library expects specific type of attachments that follow the interface. The node library allows to define attachment in simple type like string and let the library convert it

File: packages/browser/src/BacktraceBrowserRequestHandler.ts [hunk 5]
```diff
@@ -90,11 +90,11 @@ export class BacktraceBrowserRequestHandler implements BacktraceRequestHandler {
...
         }
         for (const attachment of attachments) {
-            // no support for file paths
-            if (typeof attachment === 'string') {
+            const data = attachment.get();
+            if (!data) {
                 continue;
             }
-            formData.append(`attachment_${attachment.name}`, attachment.data.toString(), attachment.name);
+            formData.append(`attachment_${attachment.name}`, data.toString(), attachment.name);
```
[the remark this answers is not in the document]
Comment (perf2711) [reply packages/browser/src/BacktraceBrowserRequestHandler.ts:97]: So are you sure that `toString()` on UInt8Array is the correct way of sending attachments? What about `Blob`s?

[the remark this answers is not in the document]
Comment (perf2711) [reply packages/browser/src/BacktraceBrowserRequestHandler.ts:97]: See also:
* https://developer.mozilla.org/en-US/docs/Web/API/File
* https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer

File: packages/node/src/builder/BacktraceClientBuilder.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (konraddysput) [reply packages/node/src/builder/BacktraceClientBuilder.ts:35]: We talked about this on the call - the type allow to pass a file attachment as a string now.

**Review by adamcronin42 — APPROVED** [review]

**Review by perf2711 — APPROVED** [review]
